### PR TITLE
Receive GRPC metadata from HTTP headers.

### PR DESCRIPTION
**Pull request by crast**
* Add support for Grpc-Metadata-(varname) headers
* Modify codegen to do metadata annotation

Also somewhat related to #4 in that it starts the pathway towards supporting authentication headers.

### CODE REVIEW

File: runtime/context.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package runtime
+
+import (
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/metadata"
```
Comment (yugui) [inline runtime/context.go]: Thank you for your contribution. 
Would you mind fixing some style issues?

Organize imports in groups.
https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Imports

File: runtime/context.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"net/http"
+	"strings"
+)
+
+const metadataHeaderPrefix = "X-Grpc-Metadata-"
+
+/*
+Add context information such as metadata from the request.
```
Comment (yugui) [inline runtime/context.go]: http://golang.org/doc/effective_go.html#commentary

The comment should be a complete sentence and should begin with "AnnotateContext adds ...".

File: runtime/context_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package runtime_test
+
+import (
+	"github.com/gengo/grpc-gateway/runtime"
+	"google.golang.org/grpc/metadata"
+
+	"golang.org/x/net/context"
```
Comment (yugui) [inline runtime/context_test.go]: Standard library packages should be the first group of imports.
https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Imports

Also `"golang.org/x/net/context"` is not a standard package (yet).

File: runtime/context_test.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+func TestAnnotateContext(t *testing.T) {
+	ctx := context.Background()
+
+	request, _ := http.NewRequest("GET", "http://localhost", nil)
+	request.Header = http.Header{}
+	annotated := runtime.AnnotateContext(ctx, request)
+	if annotated != ctx {
+		t.Errorf("Annotation did not return proper context: %v != %v", ctx, annotated)
```
Comment (yugui) [inline runtime/context_test.go]: Follow the standard convention of error message format.
https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Useful_Test_Failures

File: runtime/context_test.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	}
+	request.Header.Add("X-Grpc-Metadata-FooBar", "Value1")
+	request.Header.Add("X-Grpc-Metadata-Foo-BAZ", "Value2")
+	annotated = runtime.AnnotateContext(ctx, request)
+	md, ok := metadata.FromContext(annotated)
+	if !ok || len(md) != 2 || md["Foobar"] != "Value1" || md["Foo-Baz"] != "Value2" {
```
Comment (yugui) [inline runtime/context_test.go]: Split the check into three checks.
1. `!ok` and length check
2. `md["Foobar"]`
3. `md["Foo-Baz"]`

**Comment by crast** [comment]
OK I'll incorporate your comments

**Comment by crast** [comment]
Also, from reading a few places ( [1](http://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions) [2](http://tools.ietf.org/html/draft-saintandre-xdash-00) )   I'm going to remove the X- prefix from the header; as it's not recommended to be used

**Comment by crast** [comment]
Updated diff

**Comment by yugui** [comment]
LGTM.  Again, thanks a lot for your contribution.

**Pull request merged**